File: Infrastructure/DataAccess/LocalSecretDataAccess.cs
using System.Net.Http.Json;
using System.Text.Json;
using Infrastructure.DTOs.Secrets;
using Infrastructure.Interfaces;
using Infrastructure.Values;
using Microsoft.Extensions.Configuration;

namespace Infrastructure.DataAccess;

public class LocalSecretDataAccess : ISecretsDataAccess
{
    private readonly HttpClient _httpClient;
    private readonly string _baseUrl;

    public LocalSecretDataAccess(IHttpClientFactory httpClientFactory, IConfiguration configuration)
    {
        _httpClient = httpClientFactory.CreateClient("InsecureClient");
        _baseUrl = configuration["LocalSecretServerUrl"] ?? throw new Exception("LocalSecretDataAccess requires a url the local server. Config's LocalSecretServerUrl cannot be null.");
    }

    public async Task<string> GetKeyValue(SecretKey secretKey)
    {
        var response = await _httpClient.GetAsync($"{_baseUrl}/secrets/{secretKey.Value}");

        if (!response.IsSuccessStatusCode)
        {
            throw new Exception($"No secret of key \"{secretKey}\" exists.");
        }

        var json = await response.Content.ReadAsStringAsync();
        var options = new JsonSerializerOptions
        {
            PropertyNameCaseInsensitive = true, // Allows matching without exact case
            PropertyNamingPolicy = JsonNamingPolicy.CamelCase // Maps camelCase JSON to PascalCase C# properties
        };

        var dto = JsonSerializer.Deserialize<ReadSecretResponseDTO>(json, options);
        if (dto is null)
        {
            throw new Exception("GetKeyValue's request succeeded but failed to deserialize the JSON into a DTO."); 
        }

        return dto.Value;
    }
}

File: backend/src/infrastructure/dataAccess/LocalSecretDataAccess.ts
import ISecretDataAccess from "infrastructure/interfaces/ISecretDataAccess";
import { TSecretKey } from "infrastructure/values/secretKeys";
import fetch from "node-fetch";
import { Agent } from "https";
import ReadSecretResponseDTO from "infrastructure/contracts/secrets/ReadSecretResponseDTO";

class LocalSecretDataAccess implements ISecretDataAccess {
    constructor(private readonly secretsServerUrl: string, private readonly agent?: Agent) {}

    async getKeyValue(key: TSecretKey): Promise<string> {
        const response = await fetch(`${this.secretsServerUrl}/secrets/${key}`, {
            agent: this.agent
        });

        if (response.status === 404) {
            throw new Error(`No secret of key "${key}" exists.`);
        }

        if (!response.ok) {
            throw new Error(`Something went wrong while trying to fetch secret.`);
        }
        
        const dto = await response.json() as ReadSecretResponseDTO;
        return dto.value;
    }
}

export default LocalSecretDataAccess;
